File: front/vue3-vant-mobile/src/api/modules/health.ts
import request from '@/utils/request'
import type { ApiResponse, PageResult } from '@/types/api'

export interface HealthProfile {
  displayName?: string
  motto?: string
  heightCm?: number | null
  targetKg?: number | null
  restingKcal?: number | null
}

export interface HealthMeal {
  id: string
  date: string
  mealType: '早餐' | '午餐' | '晚餐'
  food: string
  proteinG?: number | null
  feedback?: string | null
}

export interface HealthDaily {
  date: string
  stomachStatus?: string | null
  stomachNote?: string | null
  cyclePhase?: string | null
  cycleDay?: number | null
}

export interface HealthWeight {
  id: string
  date: string
  weightType: '晨重' | '晚重'
  kg: number
  standard: boolean
}

export interface HealthSummary {
  latestMorningKg?: number | null
  avg7MorningKg?: number | null
  targetKg?: number | null
  gapToTargetKg?: number | null
  avg30ProteinG?: number | null
}

export interface HealthMemory {
  id: string
  title: string
  detail?: string | null
}

export interface HealthTolerance {
  id: string
  name: string
  level: string
}

export interface HealthTrigger {
  id: string
  name: string
  note?: string | null
  stars: number
}

export function getHealthProfile() {
  return request.get<ApiResponse<HealthProfile>>('/health/profile')
}

export function putHealthProfile(body: HealthProfile) {
  return request.put<ApiResponse<HealthProfile>>('/health/profile', body)
}

export function listHealthMeals(date: string) {
  return request.get<ApiResponse<HealthMeal[]>>('/health/meals', { params: { date } })
}

export function upsertHealthMeal(body: Omit<HealthMeal, 'id'> & { id?: string }) {
  return request.put<ApiResponse<HealthMeal>>('/health/meals', body)
}

export function deleteHealthMeal(id: string) {
  return request.delete<ApiResponse<void>>(`/health/meals/${id}`)
}

export function getHealthDaily(date: string) {
  return request.get<ApiResponse<HealthDaily>>('/health/daily', { params: { date } })
}

export function putHealthDaily(body: HealthDaily) {
  return request.put<ApiResponse<HealthDaily>>('/health/daily', body)
}

export function listHealthWeights(params?: { from?: string, to?: string, type?: string }) {
  return request.get<ApiResponse<HealthWeight[]>>('/health/weights', { params })
}

export function createHealthWeight(body: { date: string, weightType: string, kg: number, standard?: boolean }) {
  return request.post<ApiResponse<HealthWeight>>('/health/weights', body)
}

export function deleteHealthWeight(id: string) {
  return request.delete<ApiResponse<void>>(`/health/weights/${id}`)
}

export function getHealthSummary() {
  return request.get<ApiResponse<HealthSummary>>('/health/summary')
}

export function listHealthMemories(page = 1, pageSize = 5) {
  return request.get<ApiResponse<PageResult<HealthMemory>>>('/health/memories', { params: { page, pageSize } })
}

export function createHealthMemory(body: { title: string, detail?: string }) {
  return request.post<ApiResponse<HealthMemory>>('/health/memories', body)
}

export function patchHealthMemory(id: string, body: { title?: string, detail?: string }) {
  return request.patch<ApiResponse<HealthMemory>>(`/health/memories/${id}`, body)
}

export function deleteHealthMemory(id: string) {
  return request.delete<ApiResponse<void>>(`/health/memories/${id}`)
}

export function listHealthTolerances(page = 1, pageSize = 4) {
  return request.get<ApiResponse<PageResult<HealthTolerance>>>('/health/tolerances', { params: { page, pageSize } })
}

export function createHealthTolerance(body: { name: string, level: string }) {
  return request.post<ApiResponse<HealthTolerance>>('/health/tolerances', body)
}

export function patchHealthTolerance(id: string, body: { name?: string, level?: string }) {
  return request.patch<ApiResponse<HealthTolerance>>(`/health/tolerances/${id}`, body)
}

export function deleteHealthTolerance(id: string) {
  return request.delete<ApiResponse<void>>(`/health/tolerances/${id}`)
}

export function listHealthTriggers(page = 1, pageSize = 3) {
  return request.get<ApiResponse<PageResult<HealthTrigger>>>('/health/triggers', { params: { page, pageSize } })
}

export function createHealthTrigger(body: { name: string, note?: string, stars: number }) {
  return request.post<ApiResponse<HealthTrigger>>('/health/triggers', body)
}

export function patchHealthTrigger(id: string, body: { name?: string, note?: string, stars?: number }) {
  return request.patch<ApiResponse<HealthTrigger>>(`/health/triggers/${id}`, body)
}

export function deleteHealthTrigger(id: string) {
  return request.delete<ApiResponse<void>>(`/health/triggers/${id}`)
}

File: python/agent/src/life_assistant_agent/tools/health.py
"""MCP tools for health butler operations."""

from typing import Any

from ..client import JavaClient
from ..dispatch_util import invalid_action, require

_HEALTH_ACTIONS = [
    "profile_get",
    "profile_update",
    "meal_list",
    "meal_upsert",
    "meal_delete",
    "weight_list",
    "weight_add",
    "daily_get",
    "daily_update",
    "memory_list",
    "memory_add",
    "tolerance_list",
    "tolerance_add",
    "trigger_list",
    "trigger_add",
    "summary",
]


def _profile_body(fields: dict[str, Any]) -> dict[str, Any]:
    body: dict[str, Any] = {}
    for src, dst in (
        ("display_name", "displayName"),
        ("motto", "motto"),
        ("height_cm", "heightCm"),
        ("target_kg", "targetKg"),
        ("resting_kcal", "restingKcal"),
    ):
        if fields.get(src) is not None:
            body[dst] = fields[src]
    return body


def _daily_body(fields: dict[str, Any]) -> dict[str, Any]:
    body: dict[str, Any] = {"date": fields["date"]}
    for src, dst in (
        ("stomach_status", "stomachStatus"),
        ("stomach_note", "stomachNote"),
        ("cycle_phase", "cyclePhase"),
        ("cycle_day", "cycleDay"),
    ):
        if fields.get(src) is not None:
            body[dst] = fields[src]
    if "burn_kcal" in fields:
        body["burnKcal"] = fields["burn_kcal"]
    return body


def _page_params(fields: dict[str, Any]) -> dict[str, str]:
    params: dict[str, str] = {}
    if fields.get("page") is not None:
        params["page"] = str(fields["page"])
    if fields.get("page_size") is not None:
        params["pageSize"] = str(fields["page_size"])
    return params


async def dispatch(client: JavaClient, action: str, **fields: Any) -> Any:
    if action not in _HEALTH_ACTIONS:
        return invalid_action(_HEALTH_ACTIONS)

    if action == "profile_get":
        return await client.get("/health/profile")

    if action == "profile_update":
        return await client.put("/health/profile", _profile_body(fields))

    if action == "meal_list":
        err = require(fields, "date")
        return err or await client.get("/health/meals", {"date": fields["date"]})

    if action == "meal_upsert":
        err = require(fields, "date", "meal_type", "food")
        if err:
            return err
        body: dict[str, Any] = {
            "date": fields["date"],
            "mealType": fields["meal_type"],
            "food": fields["food"],
        }
        if fields.get("protein_g") is not None:
            body["proteinG"] = fields["protein_g"]
        if fields.get("kcal") is not None:
            body["kcal"] = fields["kcal"]
        if fields.get("feedback") is not None:
            body["feedback"] = fields["feedback"]
        return await client.put("/health/meals", body)

    if action == "meal_delete":
        err = require(fields, "id")
        return err or await client.delete(f"/health/meals/{fields['id']}")

    if action == "weight_list":
        params: dict[str, str] = {}
        if fields.get("from_") is not None:
            params["from"] = fields["from_"]
        if fields.get("to") is not None:
            params["to"] = fields["to"]
        if fields.get("type_") is not None:
            params["type"] = fields["type_"]
        return await client.get("/health/weights", params or None)

    if action == "weight_add":
        err = require(fields, "date", "weight_type", "kg")
        if err:
            return err
        body = {
            "date": fields["date"],
            "weightType": fields["weight_type"],
            "kg": fields["kg"],
        }
        if fields.get("standard") is not None:
            body["standard"] = fields["standard"]
        return await client.post("/health/weights", body)

    if action == "daily_get":
        err = require(fields, "date")
        return err or await client.get("/health/daily", {"date": fields["date"]})

    if action == "daily_update":
        err = require(fields, "date")
        return err or await client.put("/health/daily", _daily_body(fields))

    if action == "memory_list":
        return await client.get("/health/memories", _page_params(fields) or None)

    if action == "memory_add":
        err = require(fields, "title")
        if err:
            return err
        body: dict[str, Any] = {"title": fields["title"]}
        if fields.get("detail") is not None:
            body["detail"] = fields["detail"]
        return await client.post("/health/memories", body)

    if action == "tolerance_list":
        return await client.get("/health/tolerances", _page_params(fields) or None)

    if action == "tolerance_add":
        err = require(fields, "name", "level")
        if err:
            return err
        return await client.post("/health/tolerances", {
            "name": fields["name"],
            "level": fields["level"],
        })

    if action == "trigger_list":
        return await client.get("/health/triggers", _page_params(fields) or None)

    if action == "trigger_add":
        err = require(fields, "name", "stars")
        if err:
            return err
        body = {"name": fields["name"], "stars": fields["stars"]}
        if fields.get("note") is not None:
            body["note"] = fields["note"]
        return await client.post("/health/triggers", body)

    if action == "summary":
        return await client.get("/health/summary")

    return invalid_action(_HEALTH_ACTIONS)
